### ISSUE TITLE
Unnecessary style recommendation: semicolon after the last statement

### ISSUE THREAD

**Issue by AlexDaniel** [perl6/doc#1477]
https://docs.perl6.org/language/syntax#Separating_Statements_with_Semicolons
> A semicolon after the final statement (or after the final statement inside a block) is optional, though it's good form to include it.

What is meant by “good” here? Why? Can't we just end the sentence on “is optional”, without giving unjustified recommendations?

**Comment by coke** [comment]
This was probably meant as more of a style suggestion: same reason you'd end a multi-line list with a comma when you don't have to, to prevent yourself from adding code at the end and getting a compile error.

But that can go in a style guide. Feel free to modify this as described.

**Comment by coke** [comment]
(Note: not OUR style guide. more like a "best practices" thing if someone wants to make one.)

**Comment by smls** [comment]
FYI, moritz once started a "Well Thought-Out Practices for Perl 6" document here:
https://github.com/moritz/perl6-wtop/blob/master/practices.pod

**Comment by AlexDaniel** [comment]
> same reason you'd end a multi-line list with a comma when you don't have to, to prevent yourself from adding code at the end and getting a compile error

It's not the same. Very often you *rely* on the returned value from the last statement. The fact that you can't add code at the end without modifying the line that returns stuff is a very good reason to always omit the semicolon after the last statement.

I've been using this style (always a trailing comma after a list, never a trailing semicolon after the last statement in a block) for a few months with a great success. There were a few times when it saved me.

Also, whether you use an explicit ``return`` or not makes no difference. With or without return, the last block is always special, no matter if you care about the returned value or not. If there is ever a well thought out style guide, I'd expect it to recommend no trailing semicolon.

**Comment by zoffixznet** [comment]
👍 on removing that sentence. I too found it off-colour when I read it a few days ago. Personally, I leave off the colon deliberately, to indicate the statement's value is used as a return value.

**Comment by AlexDaniel** [comment]
(Off-topic FWIW) What @zoffixznet just said sounds reasonable too, but I'm convinced that doing that makes this style decision less prominent when someone is reading (or contributing to) your code. I've noticed that in most cases the last line in any block (even if/else) is always somewhat special, so the amount of times I had to add a semicolon to something is probably 0. I can see how it can improve the readability if you have ``do`` in front of your if/else, but at the same time I don't see myself removing semicolons if I ever decide to slap a ``do`` on something. Strict binary style decisions make it easier for me, but YMMV.
